src/pip/_internal/req/constructors.py

```diff
@@ -57,6 +58,28 @@ def convert_extras(extras: Optional[str]) -> Set[str]:
     return get_requirement("placeholder" + extras.lower()).extras
 
 
+def _set_requirement_extras(req: Requirement, new_extras: Set[str]) -> Requirement:
+    """
+    Returns a new requirement based on the given one, with the supplied extras. If the
+    given requirement already has extras those are replaced (or dropped if no new extras
+    are given).
+    """
+    match: Optional[re.Match[str]] = re.fullmatch(
+        r"([^;\[<>~=]+)(\[[^\]]*\])?(.*)", str(req)
+    )
```

Comment:
Do we need the regex to be this complicated? I thought the extra is pretty easy to detect, isn’t all needed is to scan for `\w` and split the first non-match?

Reply:
You make a good point, if I interpret [PEP508](https://peps.python.org/pep-0508/) correctly extras (if present) will always come right after the name. I don't think `\w` suffices because [dots and dashes are allowed as well](https://peps.python.org/pep-0508/#names), but even with that modification to your proposal it should be simpler than what I used here.

Reply:
Yes, the extras come immediately after, so we can scan for the name, skip any spaces if needed, remove existing extras if needed, and add the extras right in.

Reply:
Applied suggestion in 314d7c1